okapi-core/src/main/java/org/folio/okapi/util/DepResolution.java

```diff
@@ -506,27 +506,41 @@ private static void addOrReplace(List<TenantModuleDescriptor> tml, ModuleDescrip
   }
 
   private static void upgradeLeafs(
-      ModuleDescriptor md, Map<String, ModuleDescriptor> modsAvailable,
+      Map<String, ModuleDescriptor> modsAvailable,
       Map<String, ModuleDescriptor> modsEnabled, List<TenantModuleDescriptor> tml) {
-    Iterator<ModuleDescriptor> it = modsEnabled.values().iterator();
-    while (it.hasNext()) {
-      ModuleDescriptor me = it.next();
-      if (me.equals(md)) {
-        continue;
+    while (true) {
+      List<String> ret = upgradeLeafs2(modsAvailable, modsEnabled, tml);
+      if (ret == null) { // nothing done
+        return;
+      }
+      if (!ret.isEmpty()) { // error
+        return;
       }
-      ModuleDescriptor mdTo = null;
-      for (InterfaceDescriptor prov : md.getProvidesList()) {
-        for (InterfaceDescriptor req : me.getRequiresOptionalList()) {
-          if (prov.getId().equals(req.getId()) && !prov.isCompatible(req)) {
-            mdTo = lookupAvailableForProvided(modsAvailable, me, prov, mdTo);
+      // something upgraded.. try again
+    }
+  }
```

Comment:
The list values are not needed. This can be simplified if upgradeLeafs2 returns a boolean that is true if something upgraded:

    while (upgradeLeafs2(modsAvailable, modsEnabled, tml)) {
      // something upgraded.. try again
    }

Reply:
ok